contributivity_measures.py

```diff
@@ -138,3 +136,288 @@ def not_twice_characteristic(permut):
+            small_node_list = np.array([preprocessed_node_list[i] for i in subset])
+            #compute the size of subset : ||subset||
+            size_of_S=0
+            for node in small_node_list:
+                size_of_S+=len(node.y_train)
+            beta=size_of_S/size_of_I
+            return (1-beta)*first_increments[k]+beta*last_increments[k]
+
+        #compute  the importance density
+        ### compute the renormalization constant of the importance density for all datatsets
+
+        renorms=[]
+        for k in range(n):
+            list_k=np.delete(np.arange(n),k)
+            renorm=0
+            for length_combination in range(len(list_k)+1) :
+                for subset in combinations(list_k, length_combination): # could be avoided as   prob(np.array(subset))*np.abs(approx_increment(np.array(subset),j)) is constant in the combination
+                    renorm+=prob(np.array(subset))*np.abs(approx_increment(np.array(subset),k))
+            renorms.append(renorm)
+        ### defines the importance density
+        def g(subset,k):
+            return prob(np.array(subset))*np.abs(approx_increment(np.array(subset),k))/renorms[k]
+
+        # sampling
+        t=0
+        q=norm.ppf((1-alpha)/2, loc=0, scale=1)
+        v_max=0
+        while t<100 or t<q**2 *v_max  /(sv_accuracy)**2 : # Check if the length of the confidence interval  is below the value of sv_accuracy*characteristic_all_node
+            t+=1
+            print(t)
+            if t==1:
+                contributions=np.array([np.zeros(n)])
+            else:
+                contributions=np.vstack((contributions, np.zeros(n)))
+            for k in range(n):
+                u=np.random.uniform(0,1,1)[0]
+                cumSum=0
+                list_k=np.delete(np.arange(n),k)
+                for length_combination in range(len(list_k)+1) :
+                    for subset in combinations(list_k, length_combination): # could be avoided as   prob(np.array(subset))*np.abs(approx_increment(np.array(subset),j)) is constant in the combination
+                        cumSum+=prob(np.array(subset))*np.abs(approx_increment(np.array(subset),k))
+                        if cumSum/renorms[k]>u:
+                            S=np.array(subset)
+                            break
+                    if cumSum/renorms[k]>u:
+                        break
+                SUk=np.append(S,k)
+                increment=not_twice_characteristic(SUk) - not_twice_characteristic(S)
+                contributions[t-1][k]=increment*renorms[k]/np.abs(approx_increment(np.array(S),k))
+            v_max=np.max(np.var(contributions,axis=0))
+        shap=np.mean(contributions,axis=0)
+
+    return({'sv':shap, 'std_sv': np.std(contributions,axis=0) / np.sqrt(t-1),'prop':shap/np.sum(shap), 'computed_val':char_value_dict})
+
+
+#%% compute Shapley values with the regression importance sampling method
+
```

Comment:
Here instead of doing a lienar approximation I do a quadratic one with a regression.